classes/task/send_mail_request_teacher_to_create_course.php

```diff
@@ -0,0 +1,65 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * The ad hoc task for sending a email to the teacher of a course. The course is requested from a different user
+ * and the teacher is asked for permission.
+ *
+ * @package    his_unification
+ * @copyright  2018 Nina Herrmann
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+namespace local_lsf_unification\task;
+defined('MOODLE_INTERNAL') || die();
+
+require_once($CFG->dirroot . '/user/lib.php');
+
+/**
+ * Class send_mail_request_teacher_to_create_course
+ * (Task in moodle will be retried after 1 minute automatically when they throw an exception.
+ * See: https://docs.moodle.org/dev/Task_API#Failures )
+ * @package local_lsf_unification\task
+ */
+class send_mail_request_teacher_to_create_course extends \core\task\adhoc_task {
+    /**
+     * Execute the ad-hoc task.
+     * @throws \coding_exception
+     * @throws \moodle_exception
+     */
+    public function execute() {
+        $jsondata = $this->get_custom_data();
+        $data = json_decode($jsondata, true);
```

Comment:
why the extra encode/decode to/from JSON? This makes the source harder to read than necessary.

Reply:
I guess there are some plugins that do this since although `set_custom_data()` receives an array 
`get _custom_data()` returns an stdClass Object. This Problem does not occur when encoding and decoding data with json. However, it is not really necessary in this case therefore it is fixed by 89c0beac7bc512fe59bcf27df394a30bd07994d2.

Reply:
`set_custom_data()` doesn't care, it will take objects and arrays alike... and then performs `json_encode()`. 
https://github.com/moodle/moodle/blob/7e4c4b6f1ff5c190c6ac610e9c91988c9cfeb48c/lib/classes/task/adhoc_task.php#L66

Reply:
Achso :bulb:

Reply:
finally fixed by 990afac5ef245f42f937f04f4b3ebabf8c4294de.